neuraxle/metaopt/callbacks.py

```diff
@@ -0,0 +1,250 @@
+                return True
+            if validation_scores[-self.n_epochs_without_improvement] <= validation_scores[-1] and not self.higher_score_is_better:
+                return True
+        return False
+
+
+class MetaCallback(BaseCallback):
+    def __init__(self, wrapped_callback: BaseCallback):
+        self.wrapped_callback = wrapped_callback
+
+    @abstractmethod
+    def call(
+            self,
+            trial: Trial,
+            epoch_number: int,
+            total_epochs: int,
+            input_train: DataContainer,
+            pred_train: DataContainer,
+            input_val: DataContainer,
+            pred_val: DataContainer,
+            is_finished_and_fitted: bool
+    ):
+        pass
+
+class IfBestScore(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        if trial.is_new_best_score():
+            if self.wrapped_callback.call(
+                    trial,
+                    epoch_number,
+                    total_epochs,
+                    input_train,
+                    pred_train,
+                    input_val,
+                    pred_val,
+                    is_finished_and_fitted
+            ):
+                return True
+        return False
+
+class IfLastStep(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        if epoch_number == total_epochs - 1:
+            self.wrapped_callback.call(
+                trial,
+                epoch_number,
+                total_epochs,
+                input_train,
+                pred_train,
+                input_val,
+                pred_val,
+                is_finished_and_fitted
+            )
+            return True
+        return False
+
+
+class StepSaverCallback(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        trial.save_model()
+        return False
+
+
+class CallbackList(BaseCallback):
+    def __init__(self, callbacks, print_func: Callable = None):
+        self.callbacks = callbacks
+        if print_func is None:
+            print_func = print
+        self.print_func = print_func
+
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        is_finished_and_fitted = False
+        for callback in self.callbacks:
```

Comment:
todo: will need a sort function someday perhaps.

Reply:
what does that even mean ?

Reply:
closing